romankondratev-BJS2-61907: Google Photo sync

romankondratev opened the pull request.

romankondratev committed:
BJS2-61907 feat: добавить автоматическую загрузку фотографий в Google Photos

- Создан класс `GooglePhotosAutoUploader` для обработки загрузки фотографий
- Реализован метод `onNewPhotoAdded` для добавления фотографий в очередь
- Добавлен метод `startAutoUpload` для ожидания и триггера загрузки
- Создан метод `uploadPhotos` для загрузки фотографий и очистки очереди
- Реализован основной класс `Main` для демонстрации работы загрузчика с использованием многопоточности

RuslanZinovyev commented on src/main/java/school/faang/task_61907/GooglePhotosAutoUploader.java:
```diff
@@ -0,0 +1,50 @@
+    private final List<String> photosToUpload = new ArrayList<>();
+
+    public void startAutoUpload() {
+        synchronized (lock) {
```
тут вызов надо бы обернуть в бесконечный цикл

RuslanZinovyev commented on src/main/java/school/faang/task_61907/GooglePhotosAutoUploader.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    public void onNewPhotoAdded(String photoPath) {
+        synchronized (lock) {
```
можно просто метод синхронизировать в данном случае

romankondratev replied:
Это из-за того, что это единственный метод который возвращает lock.notify();?

RuslanZinovyev commented on src/main/java/school/faang/task_61907/Main.java:
```diff
@@ -0,0 +1,48 @@
+                log.info("Executor terminated successfully or timeout occurred.");
+                executor.shutdownNow();
+            }
+        } catch (InterruptedException e) {
```
```
} catch (InterruptedException e) {
    log.error("Error waiting for executor to terminate", e);
    Thread.currentThread().interrupt(); // восстановить флаг
}
```

romankondratev replied:
А нужно восстанавливать флаг в конце выполнения программ?
То есть Main уже закончился, тут программа все равно прервется.

romankondratev committed:
BJS2-61907 fix: корректная обработка прерываний потоков в 'Main' и 'GooglePhotosAutoUploader'

- Добавлен вызов `Thread.currentThread().interrupt()` в `Main` для восстановления состояния прерванного потока
- Обновлена логика в `GooglePhotosAutoUploader` для выхода из цикла при прерывании потока во время `wait`
- Улучшена безопасность потоков и обработка ошибок

romankondratev committed:
BJS2-61907 refactor(google_photos): добавить пустую строку для улучшения читаемости кода

- Добавлена пустая строка перед методом `onNewPhotoAdded`.
- Улучшена структурная читаемость файла.

BJS2-61907 fix: корректная обработка прерываний потоков в 'Main' и 'GooglePhotosAutoUploader'

- Добавлен вызов `Thread.currentThread().interrupt()` в `Main` для восстановления состояния прерванного потока
- Обновлена логика в `GooglePhotosAutoUploader` для выхода из цикла при прерывании потока во время `wait`
- Улучшена безопасность потоков и обработка ошибок

romankondratev committed:
BJS2-61907 refactor(google_photos): добавить пустую строку для улучшения читаемости кода

- Добавлена пустая строка перед методом `onNewPhotoAdded`.
- Улучшена структурная читаемость файла.

romankondratev committed:
Merge remote-tracking branch 'origin/romankondratev-BJS2-61907' into romankondratev-BJS2-61907
